csharp/src/Ice/Communicator-ObjectAdapterFactory.cs

```diff
@@ -10,48 +10,36 @@ namespace ZeroC.Ice
+
         /// <summary>Shuts down this communicator's server functionality. This triggers the disposal of all object
         /// adapters. After this method returns, no new requests are processed. However, requests that have been started
         /// before ShutdownAsync was called might still be active until the returned task completes.</summary>
-        public async Task ShutdownAsync()
+        public Task ShutdownAsync()
         {
             lock (_mutex)
             {
-                _shutdown = true;
-                _waitForShutdownCompletionSource ??= new TaskCompletionSource<object?>();
+                _shutdownTask ??= new Lazy<Task>(() => PerformShutdownAsync());
             }
+            return _shutdownTask.Value;
 
-            try
-            {
-                // _adapters can only be updated when _shutdown is false so no need to lock _mutex.
-                await Task.WhenAll(_adapters.Select(adapter => adapter.ShutdownAsync())).ConfigureAwait(false);
-            }
-            finally
+            async Task PerformShutdownAsync()
             {
-                _waitForShutdownCompletionSource.TrySetResult(null);
-            }
-        }
-
-        /// <summary>Block the calling thread until the communicator has been shutdown. On the server side, the
-        /// operation completes once all executing operations have completed. On the client side, it completes once
-        /// <see cref="ShutdownAsync"/> has been called. A typical use of this method is to call it from the main
-        /// thread of a server, which will be completed once the shutdown process completes, and then the caller can
-        /// do some cleanup work before calling <see cref="Dispose"/> to dispose the runtime and finally exists the
-        /// application.</summary>
-        public void WaitForShutdown() => WaitForShutdownAsync().GetAwaiter().GetResult();
-
-        /// <summary>Returns a task that completes when the communicator is fully shut down: see
-        /// <see cref="ShutdownAsync"/>. A typical use of this method is to await the returned task from the main thread
-        /// of a server, and then perform some cleanup work before calling <see cref="DisposeAsync"/> on the
-        /// communicator. The application can make remote invocations using a communicator that is shut down but not
-        /// disposed.</summary>
-        public async Task WaitForShutdownAsync()
-        {
-            lock (_mutex)
-            {
-                _waitForShutdownCompletionSource ??= new TaskCompletionSource<object?>();
+                try
+                {
+                    // _adapters can only be updated when _shutdownTask is null so no need to lock _mutex.
+                    await Task.WhenAll(_adapters.Select(adapter => adapter.ShutdownAsync())).ConfigureAwait(false);
+                }
+                finally
+                {
+                    _shutdownCompleteSource.TrySetResult(null);
```

Comment:
To avoid the deadlock you ran into, we should make this the continuation is ran asynchronously.

Reply:
Fixed + added comments.